helper/communicator/config.go

```diff
@@ -52,6 +58,73 @@ type Config struct {
 	WinRMTransportDecorator func() winrm.Transporter
 }
 
+// SSHConfigFunc returns a function that can be used for the SSH communicator
+// config for connecting to the instance created over SSH using the private key
+// or password.
+func (c *Config) SSHConfigFunc() func(multistep.StateBag) (*ssh.ClientConfig, error) {
+	return func(state multistep.StateBag) (*ssh.ClientConfig, error) {
+		sshConfig := &ssh.ClientConfig{
+			User:            c.SSHUsername,
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		}
+
+		if c.SSHAgentAuth {
+			authSock := os.Getenv("SSH_AUTH_SOCK")
+			if authSock == "" {
+				return nil, fmt.Errorf("SSH_AUTH_SOCK is not set")
+			}
+
+			sshAgent, err := net.Dial("unix", authSock)
+			if err != nil {
+				return nil, fmt.Errorf("Cannot connect to SSH Agent socket %q: %s", authSock, err)
+			}
+
+			sshConfig.Auth = append(sshConfig.Auth, ssh.PublicKeysCallback(agent.NewClient(sshAgent).Signers))
+		}
+
+		var privateKeys [][]byte
+		if c.SSHPrivateKey != "" {
+			// key based auth
+			bytes, err := ioutil.ReadFile(c.SSHPrivateKey)
+			if err != nil {
+				return nil, fmt.Errorf("Error setting up SSH config: %s", err)
+			}
+			privateKeys = append(privateKeys, bytes)
+		}
+
+		// aws,alicloud,cloudstack,digitalOcean,oneAndOne,openstack,oracle & profitbricks key
+		if iKey, hasKey := state.GetOk("privateKey"); hasKey {
+			privateKeys = append(privateKeys, []byte(iKey.(string)))
+		}
+
+		// gcp key
+		if iKey, hasKey := state.GetOk("ssh_private_key"); hasKey {
+			privateKeys = append(privateKeys, []byte(iKey.(string)))
+		}
```

Comment:
Why don't you change this in GCP to also use `privateKey` and remove this.

Reply:
Yes, good question, I didn't do it in this PR to avoid snowball refactor effects + this ssh config is pretty vital to Packer Builders 😄.
I would like this PR to be just a refactor an not change any behaviour.

I plan on doing this in another PR.
I also would like to refactor those `SSHHost` calls into a different PR.
I also would like to reduce usage of the state bag here also as it seems pretty easily doable.

Baby steps ! 🙂

Reply:
If that's ok 🙂

Reply:
Sounds great!

> Baby steps

Seems like a big step..

Reply:
Yes, may be 'packed steps' is a more correct term hehe 😛